discovery_child_development/notebooks/02_explore_concepts.py

```diff
@@ -11,164 +13,302 @@
+
+# %%
+# The distribution of probabilities is bimodal
+unique_concepts_df["probability"].hist()
 
-fig1.save("outputs/figures/02_unique_clusters.html")
+# %%
+# Separate out high probability and low probability concepts for each cluster
+hdbscan_highest_prob_docs = unique_concepts_df[unique_concepts_df["probability"] > 0.5]
+len(unique_concepts_df) - len(hdbscan_highest_prob_docs)
 
-# Artificial intelligence always gets assigned to the noise cluster
-unique_concepts_df[unique_concepts_df["concept"] == "Artificial intelligence"]
+# %%
+hdbscan_docs_grouped = (
+    hdbscan_highest_prob_docs.groupby("cluster")
+    .agg({"display_name": ";".join})
+    .reset_index()
+)
 
-# ... but its parent concept, computer science, gets assigned to cluster 8
-unique_concepts_df[unique_concepts_df["concept"] == "Computer science"]
+hdbscan_docs_grouped
 
-unique_concepts_df
+# %%
+hdbscan_low_prob_docs = unique_concepts_df[unique_concepts_df["probability"] <= 0.5]
 
-# +
-# Group the data by 'cluster_minilm' and aggregate the text for each cluster
-grouped_text = (
-    unique_concepts_df.groupby("cluster")
-    .agg(
-        {
-            "concept": ", ".join,
-            #'cluster': 'size'
-        }
-    )
+hdbscan_low_prob_docs_grouped = (
+    hdbscan_low_prob_docs.groupby("cluster")
+    .agg({"display_name": ";".join})
     .reset_index()
-)  # .rename(columns={'cluster': 'count'})
-# openalex_clusters.groupby('cluster_minilm')['text'].apply(' '.join).reset_index()
+)
+
+hdbscan_low_prob_docs_grouped
 
-grouped_text.head()
+# %%
+hdbscan_docs_grouped = hdbscan_docs_grouped.rename(
+    columns={"display_name": "high_prob_concepts"}
+)
+hdbscan_low_prob_docs_grouped = hdbscan_low_prob_docs_grouped.rename(
+    columns={"display_name": "low_prob_concepts"}
+)
 
-# +
-from sklearn.feature_extraction.text import TfidfVectorizer
+# %%
+hdbscan_docs_grouped = hdbscan_docs_grouped.merge(
+    hdbscan_low_prob_docs_grouped, on="cluster", how="left"
+)
+hdbscan_docs_grouped.head()
 
-# Initialize the TF-IDF vectorizer
-vectorizer = TfidfVectorizer(max_df=0.5, stop_words="english")
+# %%
+hdbscan_docs_grouped.to_csv("hdbscan_highest_prob_concept_groups.csv")
 
-# Fit and transform the aggregated text for each cluster
-tfidf_matrix = vectorizer.fit_transform(grouped_text["concept"])
+# %% [markdown]
+# # Most representative publication per concept
+#
+# It might be useful to know for each concept, which work got the highest score for that cluster.
 
-# Extract the top terms for each cluster
-top_terms = {}
-feature_names = vectorizer.get_feature_names_out()
-for i, row in enumerate(tfidf_matrix.toarray()):
-    top_terms[grouped_text.iloc[i]["cluster"]] = [
-        feature_names[index] for index in row.argsort()[-10:][::-1]
-    ]
+# %%
+most_relevant_work_per_concept = concepts_cleaned_df.loc[
+    concepts_cleaned_df.groupby("display_name")["score"].idxmax()
```

Comment:
Neat - didn't know you can find top values like this

Reply:
Ha, me neither!